Csharp10 interpolated strings

BillWagner opened the pull request:
Fixes #25679
Also fixes #26566, since it's in the same area.

C# guide updates for interpolated string enhancements. The way the compiler processes an interpolated string now uses a pattern. You can implement this pattern to construct the resulting string from an interpolated string expression. Many APIs and types in the .NET libraries already do this.

- [Preview link: LanguageReference/Attributes/Miscellenaneous](https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/general?branch=pr-en-us-26456)
- [Preview link: LanguageReference/Tokens/StringInterpolation](https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated?branch=pr-en-us-26456)
- [Preview link: WhatsNew/Tutorials/WriteACustomStringInterpolationHandler](https://review.docs.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/interpolated-string-handler?branch=pr-en-us-26456)

BillWagner committed:
language reference updates made

BillWagner committed:
add outline

gewarren reviewed (commented):
I realize it's still in draft.

BillWagner committed:
first draft of tutorial finished

Code completed, first draft of article is done.

BillWagner committed:
Apply suggestions from code review

Co-authored-by: Genevieve Warren <24882762+gewarren@users.noreply.github.com>

BillWagner committed:
fix build warnings

BillWagner committed:
grammar updates

CamSoper reviewed (changes requested):
Just one bit of feedback. Other than that LGTM

BillWagner committed:
Update docs/csharp/whats-new/tutorials/interpolated-string-handler.md

BillWagner committed:
Add notes to what's new page

BillWagner committed:
Merge branch 'csharp10-interpolated-strings' of https://github.com/BillWagner/docs into csharp10-interpolated-strings

BillWagner committed:
Apply suggestions from code review

Co-authored-by: Fred Silberberg <fred@silberberg.xyz>

BillWagner committed:
Update tutorial per offline discussion

- Add discussion of short circuiting
- Discussing additional formatting overloads

BillWagner committed:
Fix #26566

It's in the same area, so add the necessary link

CamSoper reviewed (approved):
LGTM
